File: functions/test/team_access.test.js
const {test} = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');
const {permissions, roles, allows, canManageAccessOf, migrationProposal} = require('../team_access');
const member = role => ({accessVersion:2, organizationId:'org', ownerId:'user',
  status:'active', role, buildingScope:'selected', buildingIds:['a']});

test('every role respects organization, identity, building and status boundaries', () => {
  for (const role of Object.keys(roles)) for (const permission of permissions) {
    const m = member(role);
    assert.equal(allows(m, permission, {organizationId:'org',userId:'user',buildingId:'a'}), roles[role].includes(permission));
    for (const context of [{organizationId:'other'}, {userId:'other'}, {buildingId:'b'}, {buildingId:''}]) {
      assert.equal(allows(m, permission, context), false);
    }
    for (const status of ['invited','suspended','revoked','assignmentRequired']) {
      assert.equal(allows({...m,status}, permission), false);
    }
  }
});
test('unknown, legacy, missing and malformed roles fail closed', () => {
  for (const m of [null, {}, {...member('owner'),accessVersion:1}, member('member'), member('admin'), member('toString'), member('__proto__')]) {
    for (const p of permissions) assert.equal(allows(m,p),false);
  }
  assert.equal(allows(member('owner'),'unknown'),false);
});
test('empty scope is denied and explicit all includes future buildings', () => {
  const m = {...member('manager'),buildingIds:[]};
  assert.equal(allows(m,'manageProperty',{buildingId:'new'}),false);
  assert.equal(allows({...m,buildingScope:'all'},'manageProperty',{buildingId:'new'}),true);
  assert.equal(allows({...m,buildingScope:'bad',buildingIds:['a']},'manageProperty',{buildingId:'a'}),false);
  assert.equal(allows({...member('owner'),buildingScope:'bad'},'manageTeam'),false);
});
test('only approved overrides apply and cannot bypass scope', () => {
  const m = {...member('receptionist'), permissionOverrides:{refundPayments:true,manageTeam:true}};
  assert.equal(allows(m,'refundPayments',{buildingId:'a'}),true);
  assert.equal(allows(m,'refundPayments',{buildingId:'b'}),false);
  assert.equal(allows(m,'manageTeam'),false);
  assert.equal(allows({...member('manager'),permissionOverrides:{overridePrices:false}},'overridePrices'),false);
});
test('administrator cannot change peer or owner access', () => {
  assert.equal(canManageAccessOf(member('administrator'),'administrator'),false);
  assert.equal(canManageAccessOf(member('administrator'),'owner'),false);
  assert.equal(canManageAccessOf(member('owner'),'administrator'),true);
  assert.equal(canManageAccessOf(member('manager'),'receptionist'),false);
});
test('migration proposes roles without modifying input or promoting members', () => {
  const org = {id:'org',createdBy:'owner'};
  const members = ['owner','admin','member'].map((ownerId,i)=>({id:String(i),organizationId:'org',ownerId,role:i===2?'member':'admin',status:'active'}));
  const before = JSON.stringify(members);
  const result = migrationProposal(org,members);
  assert.deepEqual(result.proposals.map(p=>p.proposed.role),['owner','administrator',null]);
  assert.equal(result.proposals[2].proposed.status,'assignmentRequired');
  assert.equal(result.proposals[2].proposed.buildingScope,'selected');
  assert.deepEqual(result.issues,[]);
  assert.equal(JSON.stringify(members),before);
  assert.deepEqual(migrationProposal(org,members),result);
});
test('migration preserves v2 assignments and flags missing owner/duplicate accounts', () => {
  const m = {...member('receptionist'),id:'m'};
  assert.equal(migrationProposal({id:'org',createdBy:'user'},[m]).proposals[0].action,'unchanged');
  const result = migrationProposal({id:'org',createdBy:'missing'},[m,{...m,id:'duplicate'}]);
  assert.ok(result.issues.some(i=>i.reason==='missingActiveOwner'));
  assert.ok(result.issues.some(i=>i.reason==='missingOrDuplicateAccount'));
});
test('Dart and server policies expose the same permission and role names', () => {
  const source = fs.readFileSync(path.join(__dirname,'../../lib/models/team_access.dart'),'utf8');
  const names = name => source.match(new RegExp(`enum ${name} \\{([^}]+)\\}`))[1].split(',').map(s=>s.trim()).filter(Boolean);
  assert.deepEqual(names('TeamRole'),Object.keys(roles));
  assert.deepEqual(names('TeamPermission'),permissions);
});

File: test/team_access_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:phan_mem_quan_ly_can_ho/models/team_access.dart';
import 'package:phan_mem_quan_ly_can_ho/models/staff_profile.dart';

void main() {
  TeamAccess access(TeamRole role, {String status = 'active'}) =>
    TeamAccess(role: role, status: status, buildingIds: {'a'});

  test('Every role is restricted to selected buildings', () {
    for (final role in TeamRole.values) {
      final a = access(role);
      for (final p in TeamAccess.defaults(role)) {
        expect(a.allows(p, buildingId: 'a'), isTrue);
        expect(a.allows(p, buildingId: 'b'), isFalse);
        expect(a.allows(p, buildingId: ''), isFalse);
      }
    }
  });
  test('Suspended, revoked, invited and unassigned users have no permissions', () {
    for (final role in TeamRole.values) {
      for (final status in ['suspended', 'revoked', 'invited', 'assignmentRequired']) {
        for (final p in TeamPermission.values) {
          expect(access(role, status: status).allows(p), isFalse);
        }
      }
    }
  });
  test('Receptionist collects but cannot refund, export or override prices', () {
    final a = access(TeamRole.receptionist);
    expect(a.allows(TeamPermission.collectPayments, buildingId: 'a'), isTrue);
    for (final p in [TeamPermission.refundPayments, TeamPermission.overridePrices,
      TeamPermission.exportData, TeamPermission.manageTeam]) {
      expect(a.allows(p), isFalse);
    }
  });
  test('Housekeeper cannot access guests or finances', () {
    final a = access(TeamRole.housekeeper);
    expect(a.allows(TeamPermission.readAssignedTasks, buildingId: 'a'), isTrue);
    expect(a.allows(TeamPermission.readBookings), isFalse);
    expect(a.allows(TeamPermission.readFinancialReports), isFalse);
  });
  test('Override cannot grant team administration or bypass building scope', () {
    final a = TeamAccess(role: TeamRole.receptionist, status: 'active',
      buildingIds: {'a'}, overrides: {TeamPermission.manageTeam: true,
        TeamPermission.refundPayments: true});
    expect(a.allows(TeamPermission.manageTeam), isFalse);
    expect(a.allows(TeamPermission.refundPayments, buildingId: 'a'), isTrue);
    expect(a.allows(TeamPermission.refundPayments, buildingId: 'b'), isFalse);
  });
  test('Owner and administrator access-management boundaries', () {
    expect(access(TeamRole.administrator).canManageAccessOf(TeamRole.owner), isFalse);
    expect(access(TeamRole.administrator).canManageAccessOf(TeamRole.administrator), isFalse);
    expect(access(TeamRole.owner).canManageAccessOf(TeamRole.administrator), isTrue);
    expect(access(TeamRole.manager).canManageAccessOf(TeamRole.receptionist), isFalse);
  });
  test('Legacy, unknown and unversioned roles fail closed', () {
    for (final map in <Map<String, dynamic>>[
      {'role':'member','status':'active'}, {'role':'admin','status':'active'},
      {'role':'owner','status':'active'},
      {'accessVersion':2,'role':'unknown','status':'active'},
    ]) {
      for (final p in TeamPermission.values) {
        expect(TeamAccess.fromMap(map).allows(p), isFalse);
      }
    }
  });
  test('Empty selected scope is not all buildings; all scope includes future buildings', () {
    final selected = TeamAccess(role: TeamRole.manager, status: 'active');
    expect(selected.allows(TeamPermission.manageProperty, buildingId: 'future'), isFalse);
    final all = TeamAccess(role: TeamRole.manager, status: 'active', allBuildings: true);
    final restored = TeamAccess.fromMap(all.toMap());
    expect(restored.allows(TeamPermission.manageProperty, buildingId: 'future'), isTrue);
  });
  test('Malformed building scope fails closed even with listed buildings', () {
    final a = TeamAccess.fromMap({'accessVersion': 2, 'role': 'manager',
      'status': 'active', 'buildingScope': 'invalid', 'buildingIds': ['a']});
    expect(a.allows(TeamPermission.manageProperty, buildingId: 'a'), isFalse);
  });
  test('Staff round trip preserves historical identity without a login', () {
    final s = StaffProfile(id: 'staff1', organizationId: 'org', code: 'S001',
      displayName: 'Historical staff', employmentStatus: 'inactive', createdAt: DateTime(2026));
    final restored = StaffProfile.fromMap(s.id, s.toMap());
    expect(restored.id, s.id);
    expect(restored.accountId, isNull);
    expect(restored.employmentStatus, 'inactive');
    expect(restored.displayName, s.displayName);
  });
}
